### PR TITLE
Script Modules API: update the code and move it to the compat/wordpress-6.5 folder

### PR DESCRIPTION
**Pull request by cbravobernal**
## What?

Modules API has been approved in Core. But was as experimental instead of 6.5 compat folder.

- https://github.com/WordPress/wordpress-develop/pull/5818

### CODE REVIEW

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (cbravobernal) [inline lib/compat/wordpress-6.5/class-wp-script-modules.php]: This differs from Core commit. Interactivity was not working in Classic Themes.
TODO: Test in Core only.
cc: @luisherranz

File: lib/compat/wordpress-6.5/scripts-modules.php
Comment (luisherranz) [inline lib/compat/wordpress-6.5/scripts-modules.php]: We also need the `wp_enqueue_module` function names.

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: You can't add two import maps.

File: lib/compat/wordpress-6.5/scripts-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/scripts-modules.php]: Maybe we should not have the `gutenberg_enqueue_module` functions in this file, so it can be copy-pasted with new changes.

File: lib/compat/wordpress-6.5/scripts-modules.php
Comment (luisherranz) [inline lib/compat/wordpress-6.5/scripts-modules.php]: Same here. This is not in WP 6.5, so maybe it should not be here.

**Comment by luisherranz** [comment]
Let's wait until a final decision on the naming has been made to merge this.

**Comment by cbravobernal** [comment]
> Let's wait until a final decision on the naming has been made to merge this.

I was about to say the same.

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (cbravobernal) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: Should we make it conditional then? Otherwise won't work on classic themes.

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: Why do you think the WP Core version won't work on classic themes? It adds the block modules and the import map to the footer:

```php
public function add_hooks() {
  add_action( 'wp_head', array( $this, 'print_enqueued_modules' ) );
  add_action( 'wp_head', array( $this, 'print_module_preloads' ) );
  add_action( 'wp_footer', array( $this, 'print_enqueued_modules' ) );
  add_action( 'wp_footer', array( $this, 'print_module_preloads' ) );
  add_action( 'wp_footer', array( $this, 'print_import_map' ) );
}
```

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: This is the explanation:

```php
/**
 * Adds the hooks to print the import map, enqueued modules and module
 * preloads.
 *
 * It adds the actions to print the enqueued modules and module preloads to
 * both `wp_head` and `wp_footer` because in classic themes, the modules
 * used by the theme and plugins will likely be able to be printed in the
 * `head`, but the ones used by the blocks will need to be enqueued in the
 * `footer`.
 *
 * As all modules are deferred and dependencies are handled by the browser,
 * the order of the modules is not important, but it's still better to print
 * the ones that are available when the `wp_head` is rendered, so the browser
 * starts downloading those as soon as possible.
 *
 * The import map is also printed in the footer to be able to include the
 * dependencies of all the modules, including the ones printed in the footer.
 *
 * @since 6.5.0
 */
```

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: More info about the reasoning in this comment:

- https://github.com/WordPress/wordpress-develop/pull/5818#discussion_r1443043837

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I haven't tested it out, though. So it's all theoretical.

Comment (cbravobernal) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I did not try in Core only yet, but at least, in Gutenberg, having the import map after the enqueued modules fires this error:
Screenshot 2024-01-12 at 11 06 13

It seems that the import map must be before the rest of scripts. Something like:
```
public function add_hooks() {
		if ( ! wp_is_block_theme() ) {
			add_action( 'wp_footer', array( $this, 'print_import_map' ) );
		} else {
			add_action( 'wp_head', array( $this, 'print_import_map' ) );
		}
		add_action( 'wp_head', array( $this, 'print_enqueued_modules' ) );
		add_action( 'wp_head', array( $this, 'print_module_preloads' ) );
		add_action( 'wp_footer', array( $this, 'print_module_preloads' ) );
		add_action( 'wp_footer', array( $this, 'print_enqueued_modules' ) );
	}
```

File: lib/compat/wordpress-6.5/scripts-modules.php
Comment (cbravobernal) [reply lib/compat/wordpress-6.5/scripts-modules.php]: I renamed the functions with the ones with `wp_` with a declaration prevention.
Should we move them to another file then?

I guess changes should be done first in this file, and later in `wordpress-develop`. I think is good to have them in the same file than Core.

File: lib/compat/wordpress-6.5/scripts-modules.php
Comment (cbravobernal) [reply lib/compat/wordpress-6.5/scripts-modules.php]: Moved back to experimental.

**Comment by cbravobernal** [comment]
~~The PR still needs some updates in tests.~~ Done

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: You're right, this doesn't work:

```html
<!DOCTYPE html>
<html>
  <head>
    <script type="module">
      import lodash from "lodash";
      console.log(lodash.add(1, 2));
    </script>
    <script type="importmap">
      {
        "imports": {
          "lodash": "https://ga.jspm.io/npm:lodash@4.17.21/lodash.js"
        }
      }
    </script>
  </head>
  <body></body>
</html>
```

And this works:

```html
<!DOCTYPE html>
<html>
  <head>
    <script type="importmap">
      {
        "imports": {
          "lodash": "https://ga.jspm.io/npm:lodash@4.17.21/lodash.js"
        }
      }
    </script>
    <script type="module">
      import lodash from "lodash";
      console.log(lodash.add(1, 2));
    </script>
  </head>
  <body></body>
</html>
```

That's quite a surprise, taking into account that modules are always deferred.

I'll investigate to see if there's any workaround.

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: There doesn't seem to be, so we need to modify that part. Let's do it here first, and then once we've confirmed that it works, submit the patch to Core.

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I'll test if we can at least still preload them in the head to start downloading them as soon as possible.

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: Oh my, that also fails 🤦‍♂️

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: We need to move everything to the footer in classic themes.

Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I opened a Trac ticket to track this fix in Core: https://core.trac.wordpress.org/ticket/60240

**Review by luisherranz — CHANGES_REQUESTED** [review]
We still need to keep `gutenberg_register_module`, `gutenberg_enqueue_module` and `gutenberg_dequeue_module` for at least a couple of versions in Gutenberg, in case someone was already using it.

I don't know how to do deprecation notices in WordPress, but if that is possible, we should add them.

**Comment by cbravobernal** [comment]
> We still need to keep `gutenberg_register_module`, `gutenberg_enqueue_module` and `gutenberg_dequeue_module` for at least a couple of versions in Gutenberg, in case someone was already using it.
> 
> I don't know how to do deprecation notices in WordPress, but if that is possible, we should add them.

This could be an example about how to do it. I'll update it.

https://github.com/WordPress/gutenberg/blob/c454a0281598e351bd773e98d6699e4a56322f61/lib/block-supports/duotone.php#L188

Works like a charm!

Screenshot 2024-01-12 at 17 40 54
Screenshot 2024-01-12 at 17 44 27

**Comment by luisherranz** [comment]
I converted it back to a draft until a [final decision on the naming](https://github.com/WordPress/wordpress-develop/pull/5818#issuecomment-1887867378) has been made.

File: lib/experimental/modules/scripts-modules.php
Comment (luisherranz) [inline lib/experimental/modules/scripts-modules.php]: We don't need those params because they were never supported in Gutenberg.

Also, shouldn't the Gutenberg version be `17.6` (the next one)? _Not sure how that works, though._

```suggestion
function gutenberg_enqueue_module( $module_id  ) {
	_deprecated_function( __FUNCTION__, 'Gutenberg 17.6', 'wp_enqueue_module' );
	wp_modules()->enqueue( $module_id );
}
```

Comment (cbravobernal) [reply lib/experimental/modules/scripts-modules.php]: 17.6, I don't know what I was thinking about 😮‍💨

**Comment by cbravobernal** [comment]
> I converted it back to a draft until a [final decision on the naming](https://github.com/WordPress/wordpress-develop/pull/5818#issuecomment-1887867378) has been made.

True

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (westonruter) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: > That's quite a surprise, taking into account that modules are always deferred.

That is surprising. I consulted Bard about this, and it says:

> * While modules are executed in a deferred manner, the process of resolving imports happens earlier.
> * When the browser encounters a module script, it immediately starts parsing it to identify imports.
> * If it doesn't have an import map available yet, it won't be able to determine where to fetch those imports from.

That makes sense to save time so module dependencies can be loaded as soon as possible. I wasn't aware that imports start to get resolved before the execution happens. I double checked the spec and it seems to [confirm this](https://html.spec.whatwg.org/multipage/scripting.html#script-processing-src-prepare), when a module script is encountered:

> 8\. Set el's [delaying the load event](https://html.spec.whatwg.org/multipage/scripting.html#concept-script-delay-load) to true.
> [...]
> 11\. [...] [Fetch an external module script graph](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-module-script-tree) given url, settings object, options, and onComplete.
> [...]
> For performance reasons, user agents may start fetching the classic script or module graph (as defined above) as soon as the [src](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-src) attribute is set, instead, in the hope that el will be inserted into the document

Fetching the external module script graph as early as possible means that the script can execute without delay once the DOM has loaded and the deferred module script is executed. In order to properly resolve the fetching of the external module script graph, the `importmap` would have to be already consumed.

File: lib/experimental/modules/scripts-modules.php
Comment (cbravobernal) [reply lib/experimental/modules/scripts-modules.php]: If we do `wp_modules()->enqueue( $module_id );` it will require at least `$module_id` and `$src`.

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: > That is surprising.

Haha, the fact that this was also a surprise for you is a bit of a relief, thank you 😄

I guess they did it to _force_ people to do it that way because it is slightly more performant. But it kind of goes against sites that discover their dependencies during the rendering while they are already streaming the HTML because it forces them to push everything till the end. Like the heavily componentized React sites.

File: lib/experimental/modules/scripts-modules.php
Comment (luisherranz) [reply lib/experimental/modules/scripts-modules.php]: `$src` is optional in `wp_modules()->enqueue`

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (westonruter) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: This could be a performance optimization that could be applied via Core-43258: if pages are output-buffered by default, then the import map could be moved to the `head` even if it is printed in the footer.

File: lib/experimental/modules/scripts-modules.php
Comment (cbravobernal) [reply lib/experimental/modules/scripts-modules.php]: I got confused with `wp_modules()->register` 🤦

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: Ok. But for now, we don't need to do anything special here, right? We just go back to this:

```js
public function add_hooks() {
  $position = wp_is_block_theme() ? 'wp_head' : 'wp_footer';
  add_action( $position, array( $this, 'print_enqueued_modules' ) );
  add_action( $position, array( $this, 'print_module_preloads' ) );
  add_action( $position, array( $this, 'print_import_map' ) );
}
```

If you confirm it, I'll prepare the PR.

Comment (cbravobernal) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I'm afraid `print_import_map` needs to go before. I updated this PR. The code should be:
```php
public function add_hooks() {
    $position = wp_is_block_theme() ? 'wp_head' : 'wp_footer';
    add_action( $position, array( $this, 'print_import_map' ) );
    add_action( $position, array( $this, 'print_enqueued_modules' ) );
    add_action( $position, array( $this, 'print_module_preloads' ) );
}
```

File: packages/block-library/src/image/index.php
Comment (cbravobernal) [inline packages/block-library/src/image/index.php]: Should these conditionals be removed once Modules API are in Core?

Comment (luisherranz) [reply packages/block-library/src/image/index.php]: Let's wait until [we've changed Webpack's Interactivity API](https://github.com/WordPress/wordpress-develop/blob/ba40e2809312337983b947fd365d037a281cd087/tools/webpack/blocks.js) to generate modules in Core.

File: lib/experimental/modules/scripts-modules.php
Comment (sirreal) [inline lib/experimental/modules/scripts-modules.php]: This should be `$module_id` matching the parameter name.

File: lib/experimental/modules/scripts-modules.php
Comment (sirreal) [inline lib/experimental/modules/scripts-modules.php]: This should be `$module_id` matching the parameter name.

File: lib/experimental/modules/scripts-modules.php
Comment (sirreal) [inline lib/experimental/modules/scripts-modules.php]: This should be `$module_id` matching the parameter name.

Comment (luisherranz) [reply lib/experimental/modules/scripts-modules.php]: Fixed in https://github.com/WordPress/gutenberg/pull/57778/commits/d9af702471ee7adbacb1fcdde53ba40d0c2cf45c.

File: lib/experimental/modules/scripts-modules.php
Comment (luisherranz) [reply lib/experimental/modules/scripts-modules.php]: Fixed in https://github.com/WordPress/gutenberg/pull/57778/commits/d9af702471ee7adbacb1fcdde53ba40d0c2cf45c.

File: lib/experimental/modules/scripts-modules.php
Comment (luisherranz) [reply lib/experimental/modules/scripts-modules.php]: Fixed in https://github.com/WordPress/gutenberg/pull/57778/commits/d9af702471ee7adbacb1fcdde53ba40d0c2cf45c.

File: lib/compat/wordpress-6.5/class-wp-script-modules.php
Comment (luisherranz) [reply lib/compat/wordpress-6.5/class-wp-script-modules.php]: I opened a PR in WordPress Core and added the code to this PR.

- https://github.com/WordPress/wordpress-develop/pull/5931

File: packages/block-library/src/image/index.php
Comment (luisherranz) [reply packages/block-library/src/image/index.php]: I've done this now that there is a PR for WP 6.5 where we can make the Webpack changes:

- https://github.com/WordPress/wordpress-develop/pull/5922

I'll make the necessary Webpack changes there.

**Comment by luisherranz** [comment]
Codewise, this should be mostly ready, but the GB CI is saying: 

```Fatal error: Cannot declare class WP_Script_Modules, because the name is already in use in /var/www/html/wp-content/plugins/gutenberg/lib/compat/wordpress-6.5/class-wp-script-modules.php on line 20``` 

and I cannot reproduce it locally.

Any ideas?

**Comment by youknowriad** [comment]
@luisherranz The class is already declared on Core (CI uses latest WordPress trunk)

**Comment by sirreal** [comment]
> and I cannot reproduce it locally.
> 
> Any ideas?

I was able to reproduce locally using wp-env with this PR checked out and using latest wordpress-develop for core _and using PHP 7.0_. This did not have any problem for me using the default PHP version.

I had a `.wp-env.override.json` like this:

```json
{
	"core": "../../wordpress-develop/build",
	"phpVersion": "7.0"
}
```

---

I pushed a fix, feel free to do what you like with it. I suspect that in PHP 7 it's parsing the entire file so it sees the class name even though it looks like there's an early return.

**Comment by luisherranz** [comment]
Thanks @sirreal!

> I suspect that in PHP 7 it's parsing the entire file so it sees the class name even though it looks like there's an early return.

Any clue why returning early was not working on this file, but it seems to work fine [in other files](https://github.com/WordPress/gutenberg/blob/80b25d3adce2f8ac1d8327001a015c5bfcec4667/lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php#L9-L11)?

```php
if ( class_exists( 'WP_Script_Modules' ) ) {
	return;
}
```

**Review by luisherranz — APPROVED** [review]
Ok, this should be ready to go, but the e2e tests are failing because this hasn't been committed to WordPress Core yet:

- https://github.com/WordPress/wordpress-develop/pull/5931

We have to wait until that happens to be able to merge.

**Comment by luisherranz** [comment]
Now that https://github.com/WordPress/wordpress-develop/pull/5931 has been committed to Core (thanks @youknowriad!) the tests pass and we can merge this.

@c4rl0sbr4v0 can you take a final look before I merge?

**Comment by sirreal** [comment]
> Thanks @sirreal!
> 
> Any clue why returning early was not working on this file, but it seems to work fine [in other files](https://github.com/WordPress/gutenberg/blob/80b25d3adce2f8ac1d8327001a015c5bfcec4667/lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php#L9-L11)?

Not really and that worries me 😅 I'd like to do some more experimentation but can't invest the time right now.

All versions of PHP seem to see the class in the same file when there's an early return. The safest thing would be to avoid that pattern completely. [See this example](https://3v4l.org/UH503):

```php
<?php
if ( class_exists( 'A' ) ) {
	echo "EXISTS A\n";
	return;
} else {
	echo "NO EXISTS A\n";
}
class A {}
```

This prints "EXISTS A" in every PHP version. I'm not sure why the "early return" ever works without errors.

[From the `include` docs](https://www.php.net/manual/en/function.include.php):

> The documentation below also applies to [require](https://www.php.net/manual/en/function.require.php).
> …
> If there are functions defined in the included file, they can be used in the main file independent if they are before [return](https://www.php.net/manual/en/function.return.php) or after. If the file is included twice, PHP will raise a fatal error because the functions were already declared. It is recommended to use [include_once](https://www.php.net/manual/en/function.include-once.php) instead of checking if the file was already included and conditionally return inside the included file.

That mentions _functions_ but I assume that also applies to _classes_, but 🤷 it's not very clear.

**Comment by cbravobernal** [comment]
I removed some "we" in the docs, as is a common review in WordPress Core. Nothing special, apart from that, it's OK to merge this once tests are done.

**Comment by cbravobernal** [comment]
> Any clue why returning early was not working on this file, but it seems to work fine [in other files](https://github.com/WordPress/gutenberg/blob/80b25d3adce2f8ac1d8327001a015c5bfcec4667/lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php#L9-L11)?

I've seen in other parts of the project that contributors add that bail out in both `class-wp-script-modules.php` and `load.php` file.

**Comment by luisherranz** [comment]
> I'm not sure why the "early return" ever works without errors.

Yep. I still see it working in other places just fine 🙂

- [WP_Font_Face](https://github.com/WordPress/gutenberg/blob/8ef546f276b6583a6ec25a7c41dc02d9358beaa2/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face.php#L12)
- [WP_HTML_Processor](https://github.com/WordPress/gutenberg/blob/8ef546f276b6583a6ec25a7c41dc02d9358beaa2/lib/compat/wordpress-6.4/html-api/class-wp-html-processor.php#L10)
- [WP_Style_Engine](https://github.com/WordPress/gutenberg/blob/8ef546f276b6583a6ec25a7c41dc02d9358beaa2/packages/style-engine/class-wp-style-engine.php#L10)

Something must be going on, but 🤷‍♂️

**Comment by sirreal** [comment]
> That mentions _functions_ but I assume that also applies to _classes_, but 🤷 it's not very clear.

[Nope, it seems like it's just functions that are problematic](https://3v4l.org/ucrGP). I'm really not sure what's going on with that one case in certain PHP versions.

File: packages/block-library/src/file/index.php
Comment (youknowriad) [inline packages/block-library/src/file/index.php:68]: Hey folks, this is breaking the package update in Core. The block library php files are copied as is into WordPress Core. Meaning we should ensure they work there without changes.

 These Gutenberg specific function calls are there for not something we should be using. I see that the previous versions had these being the IS_GUTENBERG_PLUGIN to solve that.

Comment (luisherranz) [reply packages/block-library/src/file/index.php:68]: Opsss. Our bad. I'll open a PR right away!

Comment (luisherranz) [reply packages/block-library/src/file/index.php:68]: Do you want us to release a 17.6.1 RC with the fix?

cc: @c4rl0sbr4v0

Comment (cbravobernal) [reply packages/block-library/src/file/index.php:68]: We will do a RC2. We don't need a 17.6.1 yet, as 17.6 won't be released until next week 😅

Comment (luisherranz) [reply packages/block-library/src/file/index.php:68]: There you go: 

- https://github.com/WordPress/gutenberg/pull/58242

**Comment by getdave** [comment]
I noticed this PR was merged after I raised the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) and has changed PHP files that _may_ need backporting to WP Core.

Please forgive the ping, but I marked as `Needs PHP backport` and also added to the tracking Issue.

**Comment by luisherranz** [comment]
This has already been backported in all the PRs related to the [Script Modules API](https://core.trac.wordpress.org/ticket/56313). (I reported it yesterday [here](https://github.com/WordPress/gutenberg/issues/57959#issuecomment-1910002617) but I forgot to add the backport label here, sorry 🙏).

**Comment by luisherranz** [comment]
I removed the `Needs PHP backport` label and added the `Backported to WP Core` one. @getdave let me know if that's fine.

**Comment by sirreal** [comment]
There's an issue about the `class_exists` errors: https://github.com/WordPress/gutenberg/issues/58467#issuecomment-1929390175

**Comment by getdave** [comment]
Thank you for the detailed updates @luisherranz

**Pull request merged**
